module/zfs/arc.c

```diff
@@ -1866,13 +1868,16 @@ arc_buf_try_copy_decompressed_data(arc_buf_t *buf)
 			copied = B_TRUE;
 			break;
 		}
+		b_freeze_cksum = from->b_hdr->b_l1hdr.b_freeze_cksum;
+		cnt++;
 	}
 
 	/*
 	 * There were no decompressed bufs, so there should not be a
 	 * checksum on the hdr either.
 	 */
-	EQUIV(!copied, hdr->b_l1hdr.b_freeze_cksum == NULL);
+	if (cnt > 0)
+		EQUIV(!copied, b_freeze_cksum == NULL);
```

Comment:
@ikozhukhov given that `b_l1hdr.b_freeze_cksum` can only be set when `ZFS_DEBUG_MODIFY` is enabled (see `arc_cksum_compute()`.  I think this `EQUIV` is only valid when `ZFS_DEBUG_MODIFY` is set, and we should instead wrap it with a conditional that checks for the flag.  For the same reason, the `ASSERT` after the call to `arc_buf_try_copy_decompressed_data()`  in `arc_buf_fill()` looks wrong to me as well.

My suspicion is we've never seen this in testing because by default for debug builds we enable `ZFS_DEBUG_MODIFY`.  It would be great to get @grwilson's thoughts on this, since I may be misunderstanding the intent of `b_freeze_cksum` here.

Reply:
Yes, `b_freeze_cksum` should only have data when `ZFS_DEBUG_MODIFY` is enabled. So if I understand the original problem then wrapping the `EQUIV` with `if (zfs_flags & ZFS_DEBUG_MODIFY)` would fix the problem.